service/history/taskProcessingJob.go

```diff
@@ -0,0 +1,646 @@
+		visibilityTimestamp: t.ReadLevel().visibilityTimestamp.Add(t.config.lookAheadTimeDuration),
+	}
+	if compareTaskKeyLess(&t.maxReadLevel, &newMaxReadLevel) {
+		newMaxReadLevel = t.MaxReadLevel()
+	}
+
+	// create a job for each new level
+	newLevelToDomainIDs := make(map[int][]string)
+	for domainID, level := range domainIDToNewJobLevel {
+		newLevelToDomainIDs[level] = append(newLevelToDomainIDs[level], domainID)
+	}
+	for level, outstandingTasks := range newJobsOutstandingTasks {
+		newJob := &taskProcessingJobImpl{
+			jobLevel:         level,
+			ackLevel:         t.ackLevel,
+			readLevel:        t.readLevel,
+			maxReadLevel:     newMaxReadLevel,
+			outstandingTasks: outstandingTasks,
+			domainFilter:     newDomainFilter(newLevelToDomainIDs[level], false),
+			config:           t.config,
+			logger:           t.logger,
+			metricsClient:    t.metricsClient,
+		}
+		newJobs = append(newJobs, newJob)
+	}
+
+	// the job remaining on the current level should also be split into two parts
+	// one from ack level to read level + look ahead
+	excludedDomains := []string{}
+	for domainID := range domainIDToNewJobLevel {
+		excludedDomains = append(excludedDomains, domainID)
+	}
+	newJob := &taskProcessingJobImpl{
+		jobLevel:         t.jobLevel,
+		ackLevel:         t.ackLevel,
+		readLevel:        t.readLevel,
+		maxReadLevel:     newMaxReadLevel,
+		outstandingTasks: remainingTasks,
+		domainFilter:     t.domainFilter.exclude(excludedDomains),
+		config:           t.config,
+		logger:           t.logger,
+		metricsClient:    t.metricsClient,
+	}
+	newJobs = append(newJobs, newJob)
+
+	// the other one from read level + look ahead to max read level
+	if compareTaskKeyLess(&newMaxReadLevel, &t.maxReadLevel) {
+		newJob := newTaskProcessingJob(
+			t.jobLevel,
+			newMaxReadLevel,
+			t.maxReadLevel,
+			t.domainFilter.copy(),
+			t.config,
+			t.logger,
+			t.metricsClient,
+		)
+		newJobs = append(newJobs, newJob)
+	}
+
+	return newJobs
+}
+
+func (t *taskProcessingJobImpl) Merge(incomingJob taskProcessingJob) []taskProcessingJob {
+	incomingJobImpl, ok := incomingJob.(*taskProcessingJobImpl)
+	if !ok {
+		panic("unknown task processsing job type")
+	}
+
+	if t.jobLevel != incomingJobImpl.jobLevel {
+		panic("trying to merge job with different job level")
+	}
+
+	if !compareTaskKeyLess(&incomingJobImpl.ackLevel, &t.maxReadLevel) {
+		return nil
+	}
+
+	if !compareTaskKeyLess(&t.ackLevel, &incomingJobImpl.maxReadLevel) {
+		return nil
+	}
+
+	generationFunc := []func(*taskProcessingJobImpl, *taskProcessingJobImpl) *taskProcessingJobImpl{
+		// if exists, generate the job for the first non-overlapping part (whose range is before the overlapping part)
+		generateMergedPrefixJob,
+		// generate the job for the overlapping part
+		generateMergedJob,
+		// if exists, generate the job for the second non-overlapping part (whose ranger is after the overlapping part)
+		generateMergedSuffixJob,
+	}
```

Comment:
if the content of this anonymous function does not change, make it a real function?